client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
+				TaskName:     taskName,
+				IdentityName: wid.Name,
+			})
+			if token == nil {
+				// Missing a signature, treat this case as already expired so we get a
+				// token ASAP
+				m.logger.Debug("missing token for identity", "identity", wid.Name)
+				renewNow = true
+				continue
+			}
+
+			if token.Expiration.Before(minExp) {
+				minExp = token.Expiration
+			}
+		}
+	}
+
+	if len(reqs) == 0 {
+		m.logger.Trace("no workload identities expire")
+		return
+	}
+
+	var wait time.Duration
+	if !renewNow {
+		wait = helper.ExpiryToRenewTime(minExp, time.Now, m.minWait)
+	}
+
+	timer, timerStop := helper.NewStoppedTimer()
+	defer timerStop()
+
+	var retry uint64
+
+	for {
+		// we need to handle stopCtx.Err() and manually stop the subscribers
+		if err := m.stopCtx.Err(); err != nil {
+			//TODO Close watchers
+			return
+		}
+
+		m.logger.Debug("waiting to renew identities", "num", len(reqs), "wait", wait)
+		timer.Reset(wait)
+		select {
+		case <-timer.C:
+			m.logger.Trace("getting new signed identities", "num", len(reqs))
+		case <-m.stopCtx.Done():
+			//TODO Close watchers
+			return
+		}
+
+		// Renew all tokens together since its cheap
+		tokens, err := m.signer.SignIdentities(m.minIndex, reqs)
+		if err != nil {
+			retry++
+			wait = helper.Backoff(m.minWait, time.Hour, retry) + helper.RandomStagger(m.minWait)
+			m.logger.Error("error renewing workload identities", "error", err, "next", wait)
+			continue
+		}
+
+		if len(tokens) == 0 {
+			retry++
+			wait = helper.Backoff(m.minWait, time.Hour, retry) + helper.RandomStagger(m.minWait)
+			m.logger.Error("error renewing workload identities", "error", "no tokens", "next", wait)
+			continue
+		}
+
+		// Reset next expiration time
+		minExp = time.Time{}
+
+		for _, token := range tokens {
+			id := cstructs.TaskIdentity{
+				TaskName:     token.TaskName,
+				IdentityName: token.IdentityName,
+			}
+
+			// Set for getters
+			m.lastTokenLock.Lock()
+			m.lastToken[id] = token
+			m.lastTokenLock.Unlock()
+
+			// Send to watchers
+			m.send(id, token)
+
+			// Set next expiration time
+			if minExp.IsZero() {
+				minExp = token.Expiration
+			} else if token.Expiration.Before(minExp) {
+				minExp = token.Expiration
+			}
+		}
+
+		// Success! Set next renewal and reset retries
+		wait = helper.ExpiryToRenewTime(minExp, time.Now, m.minWait)
+		retry = 0
+	}
+}
+
+func (m *WIDMgr) send(id cstructs.TaskIdentity, token *structs.SignedWorkloadIdentity) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	c, ok := m.watchers[id]
+	if !ok {
+		// No watchers
+		return
 	}
 
-	if exp, act := len(reply.SignedIdentities), len(req); exp != act {
-		return nil, fmt.Errorf("expected %d signed identities but received %d", exp, act)
+	// Pop any unreceived tokens
+	select {
+	case <-c:
+	default:
 	}
 
-	return reply.SignedIdentities, nil
+	// Send new token, should never block since this is the only sender and watchersLock is held
+	c <- token
```

Comment:
I mentioned earlier, but I think this blocks if callers fail to call the `stop` function, which would block the entire `renew` loop. Would it be safe to have a `select` and `default` here as well?

Reply:
good idea, addressed in ea9635d23d20b1e6d4d9b6967ce4a3f653ac5256

Reply:
Oooh...sorry, I misread the code. Looking at it again I understand the comment now. 

Since the channel is buffered, and we drain the channel right before `c <- token` happens, this will never block: either the watcher will consume the channel or `token` will be placed in the (now empty) channel buffer.

This also covers the buffer size `1` in https://github.com/hashicorp/nomad/pull/18431#discussion_r1330553116.

So I think the original code was fine 😅